src/DependencyInjection/DoctrineMongoDBExtension.php

```
@@ -484,6 +514,54 @@ private function normalizeDriverOptions(array $connection): array
         return $driverOptions;
     }
 
+    /**
+     * Prepare the auto encryption configuration for the connection.
+     *
+     * @param array<string, mixed> $autoEncryption The AutoEncryption configuration of a connection
+     * @param string               $defaultDB      The default database name
+     *
+     * @return array<string, mixed>
+     */
+    private function normalizeAutoEncryption(array $autoEncryption, string $defaultDB): array
+    {
+        if (! isset($autoEncryption['kmsProvider']['type'])) {
+            throw new InvalidArgumentException('The "kmsProvider" option must contain a "type" key.');
+        }
+
+        $provider     = $autoEncryption['kmsProvider']['type'];
+        $providerOpts = array_diff_key($autoEncryption['kmsProvider'], ['type' => true]);
+        // To use "Automatic Credentials", the provider options must be an empty document.
+        // Fix the empty array to an empty stdClass object, as the driver expects it.
+        if ($providerOpts === []) {
+            $providerOpts = new Definition('stdClass');
+        }
+
+        $autoEncryption['kmsProviders'] = [$provider => $providerOpts];
```

Why does this assign a `kmsProviders` key, which is later removed by the `array_diff_key()` in `loadDocumentManager()` (when calling `normalizeAutoEncryption()`)?

In the calling context, `$authEncryption['kmsProvider']` still gets used to queue up a call to `setKmsProvider()`, but it doesn't look like any of the normalization treatment above will affect that.

Please let me know if I'm overlooking something.

---

~Good catch, we don't need to set the `kmsProviders` here.~

The method `normalizeAutoEncryption` is also used to generate `driverOptions` that are injected into the client constructor. We cannot use the `Configuration` class to get the driver options because the `Client` connection is independent from the document manager and its configuration.

---

In that case, shouldn't the `normalizeDriverOptions` calling context use its own `array_diff_key()` to unset the `kmsProvider` key specific to ODM? Otherwise, I think you're passing an unrecognized option to PHPLIB and, ultimately, PHPC. It'd be silently ignored now but could become a possible low-level log message down the line.

---

I had the idea that we could use the `ODMConfiguration` class at build time to generate the `driverOptions`. But that would not work very well with environment variables. And would increase the required version of `doctrine/mongodb-odm`.